onadata/libs/utils/logger_tools.py

```diff
@@ -1468,3 +1469,54 @@ def _exec_cached_elist_counter_commit_failover() -> None:
             )
             report_exception(subject, msg)
             cache.set(ELIST_FAILOVER_REPORT_SENT, "sent", 86400)
+
+
+def delete_xform_submissions(
+    xform: XForm,
+    deleted_by: User,
+    instance_ids: list[int] | None = None,
+    soft_delete: bool = True,
+) -> None:
+    """ "Delete subset or all submissions of an XForm
+
+    :param xform: XForm object
+    :param deleted_by: User initiating the delete
+    :param instance_ids: List of instance ids to delete, None to delete all
+    :param soft_delete: Flag to soft delete or hard delete
+    :return: None
+    """
+    if not soft_delete and not getattr(
+        settings, "ENABLE_SUBMISSION_PERMANENT_DELETE", False
+    ):
+        raise PermissionDenied("Hard delete is not enabled")
+
+    if instance_ids:
+        instances = xform.instances.filter(id__in=instance_ids, deleted_at__isnull=True)
+    else:
+        instances = xform.instances.filter(deleted_at__isnull=True)
+
+    if soft_delete:
+        now = timezone.now()
+        instances.update(deleted_at=now, date_modified=now, deleted_by=deleted_by)
+    else:
+        # Hard delete
+        instances.delete()
+
+    if instance_ids is None:
+        # Every submission has been deleted
+        xform.num_of_submissions = 0
+        xform.save(update_fields=["num_of_submissions"])
+
+    else:
+        xform.submission_count(force_update=True)
```

Comment:
Would this `xform.submission_count(force_update=True)` not accomplish the same task regardless of the method of deletion taken?

```suggestion
    xform.submission_count(force_update=True)
```

Reply:
Noted.